File: go/test/connection_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/codatplatform-sdk/go"
	"github.com/voxgig-sdk/codatplatform-sdk/go/core"

	vs "github.com/voxgig-sdk/codatplatform-sdk/go/utility/struct"
)

func TestConnectionEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Connection(nil)
		if ent == nil {
			t.Fatal("expected non-nil ConnectionEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"connection": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Connection(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Connection(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := connectionBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "connection." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set CODATPLATFORM_TEST_CONNECTION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		connectionRef01Ent := client.Connection(nil)
		connectionRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "connection"}, setup.data), "connection_ref01"))
		connectionRef01Data["company_id"] = setup.idmap["company01"]

		connectionRef01DataResult, err := connectionRef01Ent.Create(connectionRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		connectionRef01Data = core.ToMapAny(entityData(connectionRef01DataResult))
		if connectionRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if connectionRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		connectionRef01Match := map[string]any{
			"company_id": setup.idmap["company01"],
		}

		connectionRef01ListResult, err := connectionRef01Ent.List(connectionRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		connectionRef01List, connectionRef01ListOk := connectionRef01ListResult.([]any)
		if !connectionRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", connectionRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(connectionRef01List), map[string]any{"id": connectionRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		connectionRef01DataUp0Up := map[string]any{
			"id": connectionRef01Data["id"],
			"company_id": setup.idmap["company_id"],
		}

		connectionRef01MarkdefUp0Name := "created"
		connectionRef01MarkdefUp0Value := fmt.Sprintf("Mark01-connection_ref01_%d", setup.now)
		connectionRef01DataUp0Up[connectionRef01MarkdefUp0Name] = connectionRef01MarkdefUp0Value

		connectionRef01ResdataUp0Result, err := connectionRef01Ent.Update(connectionRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		connectionRef01ResdataUp0 := core.ToMapAny(entityData(connectionRef01ResdataUp0Result))
		if connectionRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if connectionRef01ResdataUp0["id"] != connectionRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if connectionRef01ResdataUp0[connectionRef01MarkdefUp0Name] != connectionRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", connectionRef01MarkdefUp0Name, connectionRef01ResdataUp0[connectionRef01MarkdefUp0Name])
		}

		// LOAD
		connectionRef01MatchDt0 := map[string]any{
			"id": connectionRef01Data["id"],
		}
		connectionRef01DataDt0Loaded, err := connectionRef01Ent.Load(connectionRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		connectionRef01DataDt0LoadResult := core.ToMapAny(entityData(connectionRef01DataDt0Loaded))
		if connectionRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if connectionRef01DataDt0LoadResult["id"] != connectionRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		connectionRef01MatchRm0 := map[string]any{
			"id": connectionRef01Data["id"],
		}
		_, err = connectionRef01Ent.Remove(connectionRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		connectionRef01MatchRt0 := map[string]any{
			"company_id": setup.idmap["company01"],
		}

		connectionRef01ListRt0Result, err := connectionRef01Ent.List(connectionRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		connectionRef01ListRt0, connectionRef01ListRt0Ok := connectionRef01ListRt0Result.([]any)
		if !connectionRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", connectionRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(connectionRef01ListRt0), map[string]any{"id": connectionRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func connectionBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "connection", "ConnectionTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read connection test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse connection test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"connection01", "connection02", "connection03", "company01", "company02", "company03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("CODATPLATFORM_TEST_CONNECTION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"CODATPLATFORM_TEST_CONNECTION_ENTID": idmap,
		"CODATPLATFORM_TEST_LIVE":      "FALSE",
		"CODATPLATFORM_TEST_EXPLAIN":   "FALSE",
		"CODATPLATFORM_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["CODATPLATFORM_TEST_CONNECTION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add company_id alias for update test.
	if idmapResolved["company_id"] == nil {
		idmapResolved["company_id"] = idmapResolved["company01"]
	}

	if env["CODATPLATFORM_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["CODATPLATFORM_APIKEY"],
			},
			extra,
		})
		client = sdk.NewCodatplatformSDK(core.ToMapAny(mergedOpts))
	}

	live := env["CODATPLATFORM_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["CODATPLATFORM_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/ConnectionEntityTest.php
<?php
declare(strict_types=1);

// Connection entity test

require_once __DIR__ . '/../codatplatform_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class ConnectionEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = CodatplatformSDK::test(null, null);
        $ent = $testsdk->Connection(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "connection" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = CodatplatformSDK::test($seed, null);
        $seen = iterator_to_array($base->Connection(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = CodatplatformConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = CodatplatformSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Connection(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = connection_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "connection." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set CODATPLATFORM_TEST_CONNECTION_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $connection_ref01_ent = $client->Connection(null);
        $connection_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.connection"), "connection_ref01"));
        $connection_ref01_data["company_id"] = $setup["idmap"]["company01"];

        $connection_ref01_data_result = $connection_ref01_ent->create($connection_ref01_data, null);
        $connection_ref01_data = Helpers::to_map($connection_ref01_data_result);
        $this->assertNotNull($connection_ref01_data);
        $this->assertNotNull($connection_ref01_data["id"]);

        // LIST
        $connection_ref01_match = [
            "company_id" => $setup["idmap"]["company01"],
        ];

        $connection_ref01_list_result = $connection_ref01_ent->list($connection_ref01_match, null);
        $this->assertIsArray($connection_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($connection_ref01_list_result),
            ["id" => $connection_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $connection_ref01_data_up0_up = [
            "id" => $connection_ref01_data["id"],
            "company_id" => $setup["idmap"]["company_id"],
        ];

        $connection_ref01_markdef_up0_name = "created";
        $connection_ref01_markdef_up0_value = "Mark01-connection_ref01_" . $setup["now"];
        $connection_ref01_data_up0_up[$connection_ref01_markdef_up0_name] = $connection_ref01_markdef_up0_value;

        $connection_ref01_resdata_up0_result = $connection_ref01_ent->update($connection_ref01_data_up0_up, null);
        $connection_ref01_resdata_up0 = Helpers::to_map($connection_ref01_resdata_up0_result);
        $this->assertNotNull($connection_ref01_resdata_up0);
        $this->assertEquals($connection_ref01_resdata_up0["id"], $connection_ref01_data_up0_up["id"]);
        $this->assertEquals($connection_ref01_resdata_up0[$connection_ref01_markdef_up0_name], $connection_ref01_markdef_up0_value);

        // LOAD
        $connection_ref01_match_dt0 = [
            "id" => $connection_ref01_data["id"],
        ];
        $connection_ref01_data_dt0_loaded = $connection_ref01_ent->load($connection_ref01_match_dt0, null);
        $connection_ref01_data_dt0_load_result = Helpers::to_map($connection_ref01_data_dt0_loaded);
        $this->assertNotNull($connection_ref01_data_dt0_load_result);
        $this->assertEquals($connection_ref01_data_dt0_load_result["id"], $connection_ref01_data["id"]);

        // REMOVE
        $connection_ref01_match_rm0 = [
            "id" => $connection_ref01_data["id"],
        ];
        $connection_ref01_ent->remove($connection_ref01_match_rm0, null);

        // LIST
        $connection_ref01_match_rt0 = [
            "company_id" => $setup["idmap"]["company01"],
        ];

        $connection_ref01_list_rt0_result = $connection_ref01_ent->list($connection_ref01_match_rt0, null);
        $this->assertIsArray($connection_ref01_list_rt0_result);

        $not_found_item = sdk_select(
            Runner::entity_list_to_data($connection_ref01_list_rt0_result),
            ["id" => $connection_ref01_data["id"]]);
        $this->assertEmpty($not_found_item);

    }
}

function connection_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/connection/ConnectionTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = CodatplatformSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["connection01", "connection02", "connection03", "company01", "company02", "company03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("CODATPLATFORM_TEST_CONNECTION_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "CODATPLATFORM_TEST_CONNECTION_ENTID" => $idmap,
        "CODATPLATFORM_TEST_LIVE" => "FALSE",
        "CODATPLATFORM_TEST_EXPLAIN" => "FALSE",
        "CODATPLATFORM_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["CODATPLATFORM_TEST_CONNECTION_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }
    if (!isset($idmap_resolved["company_id"])) {
        $idmap_resolved["company_id"] = $idmap_resolved["company01"];
    }

    if ($env["CODATPLATFORM_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["CODATPLATFORM_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new CodatplatformSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["CODATPLATFORM_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["CODATPLATFORM_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
